shared/Microsoft.Extensions.WebEncoders.Sources/WebEncoders.cs

```diff
@@ -252,137 +389,774 @@ public static string Base64UrlEncode(byte[] input, int offset, int count)
+        private static int Base64UrlDecodeCore(ReadOnlySpan<char> base64Url, Span<byte> base64Bytes, Span<byte> data)
+        {
+            UrlEncoder.UrlDecode(base64Url, base64Bytes);
+            var status = Base64.DecodeFromUtf8(base64Bytes, data, out int consumed, out int written);
+
+            if (status != OperationStatus.Done)
+            {
+                ThrowHelper.ThrowOperationNotDone(status);
+            }
+
+            return written;
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static OperationStatus Base64UrlDecodeCoreSlow(ReadOnlySpan<byte> base64Url, Span<byte> data, int base64Len, out int bytesConsumed, out int bytesWritten, bool isFinalBlock)
+        {
+#if NET461
+            var base64 = new byte[base64Len];
+            return Base64UrlDecodeCore(base64Url, base64, data, out bytesConsumed, out bytesWritten, isFinalBlock);
+#else
+            byte[] arrayToReturnToPool = null;
+            try
+            {
+                var base64 = new Span<byte>(arrayToReturnToPool = ArrayPool<byte>.Shared.Rent(base64Len), 0, base64Len);
+                return Base64UrlDecodeCore(base64Url, base64, data, out bytesConsumed, out bytesWritten, isFinalBlock);
+            }
+            finally
+            {
+                if (arrayToReturnToPool != null)
+                {
+                    ArrayPool<byte>.Shared.Return(arrayToReturnToPool);
+                }
+            }
+#endif
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static OperationStatus Base64UrlDecodeCore(ReadOnlySpan<byte> base64Url, Span<byte> base64, Span<byte> data, out int consumed, out int written, bool isFinalBlock)
+        {
+            UrlEncoder.UrlDecode(base64Url, base64, isFinalBlock);
+            var status = Base64.DecodeFromUtf8(base64, data, out consumed, out written, isFinalBlock);
+
+            if (status != OperationStatus.Done && status != OperationStatus.NeedMoreData)
+            {
+                ThrowHelper.ThrowOperationNotDone(status);
+            }
 
-            // Fix up '+' -> '-' and '/' -> '_'. Drop padding characters.
-            for (var i = outputOffset; i - outputOffset < numBase64Chars; i++)
+            // Fix bytesConsumed to match the input 'base64Url' (and not the 'base64')
+            consumed = base64Url.Length - (base64.Length - consumed);
+
+            return status;
+        }
+
+#if NETCOREAPP2_1
+        private static int Base64UrlEncodeCore(ReadOnlySpan<byte> data, Span<char> base64Url, int base64Len)
+        {
+            if (base64Len > MaxStackallocBytes / sizeof(char))
             {
-                var ch = output[i];
-                if (ch == '+')
+                return Base64UrlEncodeCoreSlow(data, base64Url, base64Len);
+            }
+
+            Span<char> base64 = stackalloc char[base64Len];
+            Convert.TryToBase64Chars(data, base64, out int written);
```

Comment:
`Convert.TryToBase64Chars` is noticeable faster than `Base64.EncodeToUtf8`. Maybe because base64-encoding and byte -> char is done in one step.

Interestingly on the decoding-methods the `Base64`-methods are faster.

Reply:
/cc @ahsonkhan

Reply:
Last I checked (back in October), the span APIs were on par or faster: https://github.com/dotnet/corefx/pull/24888


@AtsushiKan, recently did a port of the optimizations to the Convert methods: https://github.com/dotnet/coreclr/pull/17033

There shouldn't be much difference on either now (in the general, non-allocating cases). Can you share details on what type of input you are measuring performance on and maybe a stripped down repro? Otherwise, I will take a deeper look in a couple of days. Thanks for bringing it up.

Reply:
Oh, I had the wrong suspects. Sorry. When running a [stripped down sequential repo](https://gist.github.com/gfoidl/b6e918a199501f86498418ab1d072921) perf is on par, as noticed by @ahsonkhan 

In the vectorized version the difference shows up. 
```c#
if (Vector.IsHardwareAccelerated && (int*)n >= (int*)Vector<T>.Count)
```
will make the difference.

For `Base64.EncodeToUtf8` `T` will be `byte`, so on AVX2 `Vector<T>.Count` is 32.
For `Convert.TryToBase64Chars` `T` will be `char` (or to be precise `ushort`), hence `Vector<T>.Count` is 16.

For a Guid `n` is 24. So the _Convert-version_ will urlencode vectorized, whereas the _Base64-version_ will run sequential only.